src/test/ui/associated-types/defaults-in-other-trait-items.rs

```diff
@@ -0,0 +1,48 @@
+// compile-fail
+
+#![feature(associated_type_defaults)]
+
+// Associated type defaults may not be assumed inside the trait defining them.
+// ie. they only resolve to `<Self as Tr>::A`, not the actual type `()`
+trait Tr {
+    type A = ();
+
+    fn f(p: Self::A) {
+        let () = p;
+        //~^ ERROR mismatched types
+        //~| NOTE expected associated type, found ()
+        //~| NOTE expected type `<Self as Tr>::A`
+    }
+}
+
+// An impl that doesn't override the type *can* assume the default.
+impl Tr for () {
+    fn f(p: Self::A) {
+        let () = p;
+    }
+}
+
+impl Tr for u8 {
+    type A = ();
+
+    fn f(p: Self::A) {
+        let () = p;
+    }
+}
+
+trait AssocConst {
+    type Ty = u8;
+
+    // Assoc. consts also cannot assume that default types hold
+    const C: Self::Ty = 0u8;
```

Comment:
Can you extend this to checking the *values of constants* cannot be assumed to hold?

Reply:
That sounds like an orthogonal test that would need to hold for assoc. consts in general (where default values are already stable). I can certainly add that test though, but it'd have to use an assoc. const that actually type-checks, so might as well add it to the `associated-consts` directory, right?

Reply:
Sure; either works for me. I see this as testing provided things in general rather than specifically for associated types... I'm not sure what the best test organization is; perhaps there should be a folder dedicated to defaults in general... but `associated-consts` is fine if you prefer that for now.